### ISSUE TITLE
Proposal: Replace env var and file inputs with JSON on STDIN

### ISSUE THREAD

**Issue by ryanmoran** [deislabs/cnab-spec#114]
# Summary
According to the spec, installation details, image maps, parameters and credentials are all mapped on to environment variables or files that are then injected into the invocation image when an action is executed. This design, when implemented at a large scale for complex bundles, has the negative result of introducing a significant burden on the bundle author when determining and parsing the configuration inputs to their run tool.

This proposal seeks to streamline the input process for runtimes and run tools, opting to send all input, serialized as a JSON document, on STDIN when performing the action against the invocation image. The choice of JSON and STDIN, while introducing new tradeoffs will hopefully scale more gracefully as bundles increase in complexity.

Let's take some time to look at the current specification and outline its limitations.

### Environment Variables as input
The choice of environment variables as one of the mechanisms for input configuration has resulted in some undesirable limitations on the specification of parameters and credentials.

#### Limited Input Types
As described in the specification, a bundle author can map any one of their parameter or credential inputs to an environment variable. This mapping creates a logical one-to-one mapping between parameters and environment variables (ignoring for the moment that you could choose files instead). Because each environment variable within the running invocation image environment maps to a single parameter, we must limit the types of inputs that a bundle author can specify on any one parameter to those types that can cleanly serialize to an immediately consumable string format (string, int, or boolean).

Interestingly, attempting to specify other types, like object or array will result in undetermined results for each runtime. In the case of duffle, we see that specifying a parameter type for an object in our bundle.json is not validated by the runtime and results in an environment variable or file with contents that match the string-serialized representation of a Go map. You can imagine that other runtimes will have differing results.

Now, this isn't to say that passing more complex structured data into a run tool is not possible with the spec we have today. Bundle authors could choose to declare a single string type parameter into which they could pass structured data (possibly a JSON document). While this could work, we lose all of the benefits that were gained in specifying the structure and type of input configuration. By pushing all of the inputs into a single string field, we shift the burden of validating input onto the bundle author. We also lose the advantage of being able to see a complete set of configuration inputs for a bundle given just the bundle.json document.

#### `execve: Argument list too long`
Assuming we do choose to limit ourselves to simple primitive inputs (string, int, or boolean), we run into another issue with environment variables. Let's imagine that we are bundling a distributed system that consists of 10 processes that communicate over HTTPS. These processes use mutual TLS to secure their communications, requiring that when we configure each process we provide a public/private key pair and a set of CAs to use to verify the certificates of other processes. So, lets say that each of these PEM-encoded values has a size of 4kB. To provide a complete set of configuration inputs for our bundle, we will need 30 string parameters. This brings the total size of inputs for our run tool to something around 120kB. Now, on many UNIX systems today, the total limit for the length of arguments passed to exec is 128kB (configurable on a per-host basis). This leaves about 8kB for us to provide all of the other configuration for our rather complicated 10-process distributed system. If we can't fit the remaining configuration in those 8kB, then we will see something like the following from our system when we go to invoke the CNAB runtime:
```
exec user process caused "argument list too long"
```
The limit here is on the total of all arguments passed to exec. We can have 1 really long argument or very many small ones, but if they sum up to a length that is longer than the host-specified limit, the OS refuses to exec the child process. This clearly creates an upper bound on the size of configuration inputs provided via environment variables. And worse, that limit could be different on different hosts.

### Files as input
Maybe the answer is files. Files don't have such a small size constraint. However, from a security standpoint, they have their own problems. Assuming that we would be storing secret values in either credentials or parameters, there is certainly a security concern that these values are written to disk when we choose a file mapping.

Assuming that a container filesystem becomes compromised, we've just made all of our credentials easily consumable for an attacker. Add this concern to those outlined in the "Limited Input Types" section above (but rephrased to describe files), and files start also looking less than ideal.

## Another Option
To overcome the limitations outlined above, we propose we make a couple of changes to the spec that will allow configuration inputs to scale dramatically while keeping most of the benefits we already see from providing a set of well-specified parameters and credentials.

### JSON as a Serialization Format
The first change is to take all inputs (installation details, image maps, parameters, credentials, etc.) and combine them into a single JSON document that can be read by a run tool during execution. The advantage to using JSON as a serialization format is that we gain the ability to pass structured data into our run tool process. That structured data can cover all of the types of structured data inputs that are outlined in the full specification of JSON Schema. Additionally, JSON is a well-supported serialization format in most widely-used programming languages. Finding and using libraries to parse and validate a JSON document would be a minimal burden on bundle authors to also bring the advantages of a structured data format.

By choosing a structured data format, we would no longer limit ourselves to simple string, int, or boolean inputs, and expand that set of options to include object and array. This addition enables bundle authors to specify inputs, like an unbounded list of ports, for their run tool in a format that is natively understood by the runtime. The advantage for bundle authors and users would be that the requirements for this input would be documented within the bundle.json document and that any given runtime would be able to validate this more complicated input. Both of these improvements are major benefits to the specification and its users.

To make things a bit more concrete, here is a "straw man" example of the configuration structure passed to the invocation image on STDIN:
```
{
  "installation": {
    "name": "my-app",
    "action": "install",
    "revision": "01D43QHNM3PC1KHA6ND9S9H109"
  },
  "invocation_image": {
    "imageType": "docker",
    "image": "deislabs/helloworld-cnab:857f0df8ebf1ec9305a5fb105cb32d6dd65ce6f7",
    "platform": {}
  },
  "bundle": {
    "name": "helloworld",
    "version": "0.1.0",
    "description": "A short description of your bundle",
    ...
  },
  "parameters": {
    "greeting": "hello"
  }
  "credentials": {}
}
```

In the current specification, bundle authors provide installation details through environment variables CNAB_INSTALLATION and CNAB_ACTION in the run tool. In this example document, we can see all input fields for installation details, invocation image details, the entire bundle specification (including image maps), and a set of resolved parameter and credential values are in a single location. Providing all of these values in a one place would streamline the process of locating and processing inputs to the run tool for bundle authors. This brings us to the second change.

### Pass Configuration Input on STDIN
The second change proposed is that we change the mechanism whereby input data is injected into the run tool. Instead of using environment variables or files, we should inject this data into the run tool on STDIN. The advantages of using STDIN are that it bypasses the length and security concerns outlined for environment variables and files above. The length of the JSON document passed via STDIN could far surpass any limit that would be imposed by the OS for the length of arguments passed to exec. Furthermore, these input values would be read most immediately into the memory of the run tool process, further reducing the risk that a compromised container would result in the leaking of secret configuration information.

## Tradeoffs
There are obvious tradeoffs to the proposed changes outlined above. At a cursory glance, we can outline a few that have an impact on bundle authors.

1. By proposing that we change the injection mechanism for configuration inputs, we lose the immediate mapping to environment variables or files the might have benefitted some bundle authors.
2. We declare a run tool API dependency on the JSON document format.
3. We move input from forms that developers might be more comfortable with (environment variables or files) to a less commonly used form (STDIN).

However, for each of those concerns, we also see benefits.

1. For bundle authors with many complicated configuration inputs, the process of reading and processing these inputs into a run tool is considerably more streamlined.
2. We gain the advantage of well-specified, structured data that can be discovered and validated directly from the bundle.json document.
3. We remove the constraints on input size and improve the security posture of handling secrets within the invocation image environment.

Additionally, there are some tradeoffs for the specification itself and the various runtimes. While we complicate the types of parameter inputs that can be specified by a bundle author, we also reduce the complexity of parameter and credential mapping for both the specification and its runtimes. Bundle authors will be able to specify much more complicated parameter inputs, but spec documents will no longer need to include numerous field to indicate mappings onto environment variables or files within a given runtime. Additionally, runtimes themselves can remove all of the complex code that maps these inputs onto environment variables or files before invoking the run tool.

## Next Steps
In proposing this change, we hope to spark a conversation about the limitations and complexities introduced by the specification in its current form. Additionally, we are looking to hear feedback on the tradeoffs we have made in our proposal as well as possible alternatives that we did not explore. Thanks! 😄

/cc @youreddy

**Comment by carolynvs** [comment]
I have a naive question. Right now the implementations focus on docker, but in theory they should work with a virtual machine instead. Would this (passing via STDIN) work with a VM as well?

**Comment by ryanmoran** [comment]
The only VM driver example that I have seen is the [azure-vm](https://github.com/deislabs/duffle/tree/master/drivers/azure-vm) one. I believe that the Azure VM driver spins up a VM from the invocation image, opens an SSH connection to the VM, and then executes the run tool. I believe that passing data on STDIN when executing that tool would be possible.

**Comment by carolynvs** [comment]
The more I think about this the more I see this as quite disruptive of a change. Especially for tools that implement the spec, like porter. It would be quite different if STDIN was simply another option instead of a hard switchover from env vars and files which are quite useful in many cases.

I could provide some use cases and explain why taking those away would make life more difficult and hurt usability for some people writing bundles. But first, I want to double check that a hard switch to STDIN is part of the proposal and that we couldn't keep both around perhaps as a compromise?

**Comment by technosophos** [comment]
Why "replace" and not just offer as another way of doing things?

**Comment by technosophos** [comment]
Thanks for taking the time to put this together.

I would be enthusiastic about adding `STDIN` as one of the possible routes for injecting param and credential data to the run tool. It would indeed get around some of the limitations outlined above, and that change alone I believe is sufficient for addressing the more pressing concerns of security and arbitrary size limits.

I would be opposed to _replacing_ env var and file injection with STDIN because it changes the entire profile of how a CNAB bundle is authored and executed. Run tools would have to get far more complex when bundles deal with lots of parameters and credentials. If nothing else, this requires that every invocation image now understands, and be equipped to parse, JSON data (though the complexity is actually deeper, since it now requires run tools to be able to delegate credentials and parameters to other commands, which means each run tool will need to be fairly sophisticated). In general I believe we should push back at anything that makes authoring bundles more complex.

Keep in mind that one thing this proposal has not addressed is the _reason_ why we put this data in env vars and files, which is that the underlying tools we already use (like Terraform, kubectl, Ansible, etc) can directly use this data as-is. All of that is lost if we were to wholesale switch to STDIN+JSON.

On that note, in my opinion, neither your first nor second benefit is actually achieved with the proposed solution. Instead, (1) bundles are more complex (not streamlined), and (2) you add a new requirement to add a new tool to every CNAB bundle -- one that would also then have to know how to perform additional validation (this, btw, can be done now if you want. JSON actually doesn't enable the behavior, nor does it mandate the behavior). Your third benefit accrues by simply allowing STDIN as one of the choices.

I would recommend, as an illustration, looking through the examples in the example bundle repo, and trying to figure out how to replace all credentials and parameters there with the system you have proposed above.

My feeling, then, is that adding STDIN as _an option_ in addition to files and env vars is a great idea. But (a) not as a replacement, and (b) not with the stipulation that all data be passed in as JSON data.

Thanks again for taking the time and effort to put this all together. This is definitely _the time_ to talk about ideas like this.

**Comment by ryanmoran** [comment]
Thank you both, @carolynvs and @technosophos for the feedback. I don't feel that the proposal requires that we replace env vars and files. I can see specifying STDIN as an option as a viable compromise. However, @technosophos, I am unclear how you would achieve a parsable STDIN input for multiple parameters if you didn't provide that input as some sort of structured document.

For example, assuming that I need 3 parameters, `first_name`, `middle_initial`, and `last_name`, to run my application. And for each of these parameters I specified `stdin` as their destination. How would a bundle author be able to differentiate each of the parameters from one another?

**Comment by ryanmoran** [comment]
On another front for discussion, our primary concern in opening this proposal is to enable a specification for parameters that would allow bundle authors to declare a complete and well-specified set of inputs for a bundle. We have plans to develop our own runtime that would want to expose parameter inputs to the user via a well-defined RESTful API. Additionally, we would like for that API to be able to validate the input for correctness ("is the input a list of IP addresses?"). In order to achieve such a service, we will want declared support for `object` and `array` types within the parameters specification.

In [other conversations](https://github.com/deislabs/cnab-spec/pull/77#issuecomment-461563677) around the topic of adding more robust type support for parameters, it has been suggested that we could have users pass in a well-formed and complete JSON document via a `string` parameter. This would certainly work for bundles that have a small set of parameters, but as the set of parameters grows, humans would become less capable of crafting such an input correctly.

Our concern with mitigations like the one described above is that without the addition of first-class `object` and `array` types, authors are left with an impoverished grammar for defining their inputs. In that situation, authors of sufficiently complex bundles would likely start to introduce external schema documents to fill in the missing parameter specifications such that a complete and well-specified set could be provided to the run tool.

For example, if I need to specify a list of `strings`, I could define a schema external to my `bundle.json` that indicates that the `list_of_strings` parameter is a JSON-encoded array of strings. Maintaining and distributing that external schema document is a burden for bundle authors, even in the simplest of cases. For our use-cases, these external schema documents would be considerably more complex than the example given here.

With nearly every one of our use-cases, we would need to include one of these external schema documents such that our runtime would be able to validate inputs from the user. That likely means that we would be introducing some sort of extension to the spec that would make bundles developed for our runtime incompatible with the rest of the runtimes in the CNAB ecosystem. That is certainly an outcome we would like to avoid.

What are y'alls thoughts? Do we not feel like validation of inputs is an important value proposition for a runtime developed on top of this specification?

**Comment by technosophos** [comment]
I understood your previous request (#77) to be more like this: "I want to define a parameter that is described by a JSON Schema and is injected as string data into the invocation image." In that case, what the spec calls the "CNAB runtime" validates the parameter input by the following process:

1. Read the parameter definition from the bundle descriptor (`bundle.json`)
2. Load the parameter data from the user-provided source
3. Parse the data using a JSON parser
4. Apply the schema given in the parameter definition
5. Serialize the JSON object back into a string
6. Inject it as an env var or file, as specified in the bundle descriptor

Now, in _this_ case, what you get is...

- Support for "structured data" where needed
- Still support for all the basic data types
- Freedom of the bundle author to write things the way they want, including writing specialty schemata for their exact data strutctures
- Neutrality of destination (data can go into env vars, files, stdin)
- Separation of parameters and credentials

So how I was envisioning what you were asking about in #77 is more like this:

```json
{
  "parameters": {
      "backend_port": {
        "type": "integer",
        "defaultValue": 80,
        "minimum": 10,
        "maximum": 10240,
        "metadata": {
          "description": "The port that the back-end will listen on"
        },
        "destination": {
          "env": "HTTP_PORT"
        }
      },
      "list_of_strings": {
        "type": "array",
        "schema": "TBD -- maybe embedded JSON Schema, maybe $schema/$url marker...",
        "defaultValue": ["one", "two"],
        "metadata": {
          "description": "a list of strings"
        },
        "destination": {
          "file": "/etc/list_of_strings.json"
        }
      } 
    },
}
```

In the case above, the value for the parameter named `list_of_strings` would be loaded into the CNAB runtime, which would then parse the value, and then use the content of `schema` to validate the content against a schema. If validation failed, the CNAB runtime would bail out with the appropriate error. Otherwise, the JSON would be serialized (or perhaps the initially provided value would just be forwarded on) and stored in the file `/etc/list_of_strings.json`

Note that the `HTTP_PORT` could still be specified as a separate parameter (and likewise, this system would be in place for credentials as well).

I believe, if I am understanding your concerns above and in #77, that this would solve your needs when passing in large amounts of structured data, but would not require _everyone_ to do things that way.

Coupled with allowing STDIN as a destination, this could be a way of achieving exactly what was asked in the proposal, but with one caveat: The CNAB runtime does the validation, not the invocation image. If it is a requirement that the invocation image _also_ does the validation (and I can imagine cases where this might be true), then that tooling would be done by the bundle author inside of the invocation image.

Sorry for the confusion here. I should have given an example in the comments on #77.

**Comment by ryanmoran** [comment]
If I led you to believe that I wanted the validation to take place inside the invocation image, that is my mistake. While validating inputs inside an invocation image is certainly a workflow we wouldn't want to prevent, its not our main concern.

The outcome you describe of having a CNAB runtime that can validate a complete set of inputs from a user is what we hope to achieve. I believe I am starting to see a path forward becoming more clear, but I want to pose a couple clarifying questions first:

1. In your example above, if the bundle author had declared `stdin: true` as the `destination` for both parameters (`backend_port` and `list_of_strings`), what would the contents of `stdin` look like when the run tool went to read it?
2. In an aside above you state `and likewise, this system would be in place for credentials as well`. Does this mean that we would be extending all of the JSON Schema type declaration to credentials as well?
3. In the definition of the new `schema` field, are we deviating from JSON Schema as the structure of this section of the `bundle.json`?
4. Would the `type` for the `list_of_strings` parameter actually be `string` as it is simply a string that conforms to the given schema?

**Comment by simonferquel** [comment]
I have a mixed feeling about that: if we use stdin for passing parameters, we can't use it for anything else in the future (like, support for interactive installers, etc.)

I think it requires some more thinking about the implications, and maybe defer the decision for after v1.

**Comment by ryanmoran** [comment]
@simonferquel I'd love to hear more of an explanation of this "interactive installers" idea. I'd also like to hear how you feel using stdin for this purpose would preclude us from extending the specification to support new features.

**Comment by simonferquel** [comment]
@ryanmoran with the well-known actions and action-specific parameters, I have one example in mind:
If we have a well-known action for logging, with a parameter "follow", it would be handy to use stdin to signal that we want to stop following logs. 
Also, in the future, I'd like to be able to have support for interactive actions (that you would interact with using stdin/stdout).

**Comment by youreddy** [comment]
@simonferquel We could support the the logs use case you are describing by having these types of actions handled differently by the run tool. The proposal describes a JSON document that gets passed to the run tool via stdin. The logs action could consume that document and either pull useful information from it (i.e configuration options like filtering) or it could simply ignore that document and interpret the remaining content on stdin. When the user issues an interruption (ctrl + c) to stop streaming the logs, the run tool would be responsible for terminating the logs action. 

This potentially introduces some complexity in the run tool. For instance, the tool could determine that this type of interruption results in terminating the log streaming process. Or more simply background the process so that it could be reattached to again. 

In terms of the interactive actions, I could imagine an action like "debug" that allows a user to open a shell prompt on the container (like `docker exec`) and interact with the running service. Is that along the lines of what you're thinking?

**Comment by ryanmoran** [comment]
What if the proposal was that we would always pass the input configuration JSON document as described above on stdin. We could then relax the requirement of a `destination` field for parameters. The result would be that all parameter and credentials would be available as a JSON document on stdin by default. However, if a bundle author wanted to make a single parameter available as an environment variable or file, they could provide an optional `destination` field.

This would have no impact on the current set of example bundles as the functionality for destinations does not disappear. It also enables arbitrarily complex and large configurations for those bundles that need it with a limited impact on the existing spec.

What do folks think?

**Comment by simonferquel** [comment]
To avoid using stdin, we could do this like we do for image-map injection: inject the json payload at a well known location (e.g.: `/cnab/input-parameters.json`) into the container before starting it. 

It fixes both my reservation about using stdin, the original problem, and it applies the same strategy as an existing aspect of the spec.

**Comment by ryanmoran** [comment]
@simonferquel I still have reservations about the security profile of writing a file that may contain secrets to disk. In an ideal world, we wouldn't ever write secrets to disk, especially not in plain text. Do you have any ideas for how we could provide the type of functionality we describe in our proposal without also introducing a possible security concern?

**Comment by simonferquel** [comment]
Parameters should not hold sensitive data, credentialset are here for that.
Today the docker driver (which does not leverage any orchestrator feature such as secrets) copies the credentials in the invocation container in the same way. But I'd like to see a Kubernetes driver that assumes credential sets are stored in k8s secrets, and that would mount these secret in a Pod instead of copying them in the container RooFS layer. It is another subject, and it does not require any change in the spec (the spec says the runtime will make the files available at runtime, copying to the container RootFS before starting the container is just one way to achieve this, mounting secrets would be another).

**Comment by ryanmoran** [comment]
@simonferquel this definition of the differences between parameters and credentials does not align with how @technosophos is describing them in https://github.com/deislabs/cnab-spec/issues/39#issuecomment-462395894, nor does it align with the recent additions in https://github.com/deislabs/cnab-spec/pull/106.

If you read those definitions, then you come away with an understanding that parameters may contain secrets (these are the secrets that are configuration of the application) and that credentials may not (they are simply the configuration of the components within the run tool that perform the action).

**Comment by technosophos** [comment]
Just checking back in on this one. Trying to summarize where we are:

- So the current argument _against_ exposing `stdin` in the `destination` object is that it would prevent us from using STDIN for other things (like sending signals?).
- The argument against injecting this same data as files is the security argument (we might not want tokens sitting around on disk)
- There was a proposal to ALWAYS pass some data on `stdin`

I'm not too keen on always passing data on `stdin`. I still feel like that is overly opinionated about how a run tool should behave. But I think that allowing STDIN as a target is something that we should very seriously consider.

I also like the idea that came up on the call two weeks ago that if we allow STDIN as a destination, and we allow multiple things to write to STDIN, then we could conceivably send multiple JSON documents over the way some log streaming services do. I could imagine that proving pretty powerful.

Let's try to list what we lose if we allow STDIN to be a target. That might help us make a more informed decision.

**Comment by radu-matei** [comment]
We had a meeting about this topic earlier today, and for future reference, here are the notes:

Concerns:
- (Simon F.) only one STDIN -- if other features want to use this in the future, this becomes an issue
- (Matt B.) making sure runtimes support STDIN (Kubernetes?) in a clear way
- (Matt B.) limit to one or multiple STDIN messages? - we should not enforce this in the spec

Other possible approaches:
- (Simon F.) approaches that don't require STDIN and don't write unencrypted to disk:
    - create Kubernetes driver that uses secrets
    - for the Docker driver, use tmpfs volumes, without writing secrets to disk

- (Ryan M.) -- security is the main concern. Writing credentials in JSON is unacceptable, and environment variables are not enough. We have experience in using STDIN for this use case in the past:
    - really secure input for credentials
    - highly structure data that needs to be validated
    - declare a schema for the input on the parameters / credentials, and the runtime to provide validation.

Action items:
- PR on the spec for the schema validation for parameters and credentials
- explore the alternative ways (Docker in memory volumes and Kubernetes secrets)

**Comment by ryanmoran** [comment]
With the merging of #153, I think we can consider this proposal closed. We now have support within the spec for complex configurations and we have a path forward for securing any secrets that might be written out as files.
